src/foam/u2/detail/AbstractSectionedDetailView.js

```diff
@@ -102,6 +106,214 @@ foam.CLASS({
+          // triggered when `data` changes?
+          Promise.all(visibleSections.map(s => {
+            var visibleProperties = s.properties
+              .map(prop => {
+                // First filter by visibility based on controllerMode since
+                // that's cheapest.
+                if ( this.controllerMode.getMode(prop) === foam.u2.DisplayMode.HIDDEN ) return null;
+
+                // Next filter by visibility or visibilityExpression since
+                // that's cheaper than the permission check.
+                var vis = prop.visibilityExpression
+                  ? this.data.slot(prop.visibilityExpression).get()
+                  : prop.visibility;
+                if ( vis === foam.u2.Visibility.HIDDEN ) return null;
+
+                return prop;
+              })
+              .filter(prop => prop);
+
+            // Filter out the actions that user can't see.
+            var visibleActions = s.actions
+              .map(action => {
+                // First check based on isAvailable, not including permission
+                // check.
+                if ( ! action.isAvailable ) return action;
+
+                return this.data.slot(action.isAvailable).get() // We assume isAvailable is synchronous.
+                  ? action
+                  : null;
+              })
+              .filter(action => action);
+            
+            if ( visibleProperties.length + visibleActions.length === 0 ) {
+              // No need to do the permission check stuff if we already know
+              // the section is empty.
+              return;
+            }
+
+            // Now we almost have the list of properties that are visible. The
+            // last thing we need to do is filter out the ones that the user
+            // doesn't have permission to see.
+            var propertyPermissionPromises = visibleProperties.map((prop) => {
+              var propName = prop.name.toLowerCase();
+              var clsName  = prop.forClass_.substring(prop.forClass_.lastIndexOf('.') + 1).toLowerCase();
+
+              if ( ! (prop.readPermissionRequired || prop.writePermissionRequired) ) {
+                return Promise.resolve(true);
+              }
+
+              return this.auth.check(null, `${clsName}.rw.${propName}`).then((rw) => {
+                if ( rw ) return foam.u2.Visibility.RW;
+                return this.auth.check(null, `${clsName}.ro.${propName}`).then((ro) => ro ? foam.u2.Visibility.RO : foam.u2.Visibility.HIDDEN);
+              });
+            });
+
+            // Filter out actions that the user doesn't have permission to see.
+            var actionPermissionPromises = visibleActions.map(action => {
+              return Promise.all(action.availablePermissions.map(permission => {
+                return this.auth.check(null, permission);
+              }));
+            });
+
+            return Promise.all([
```

Comment:
Do these need `catch` callbacks or is it okay like this?

Reply:
Addressed by 1d8e5c759.